File: app/services/ai/attack_group_builder.rb
# frozen_string_literal: true

module Ai
  class AttackGroupBuilder < BaseAnalyzer
    SYSTEM_PROMPT = <<~PROMPT
      You are an expert at creating coordinated action plans for addressing product issues.
      Analyze the insights, themes, and ideas to create "attack groups" - bundles of related
      issues and solutions that should be tackled together for maximum impact.

      For each attack group:
      1. Create a compelling name for the initiative
      2. Write an executive summary
      3. List the insights being addressed (by index)
      4. List the recommended ideas to implement (by index)
      5. Identify the key themes this addresses (by index)
      6. Estimate combined effort and impact
      7. Suggest an execution order for the ideas
      8. Highlight dependencies and risks

      Respond in JSON format:
      {
        "attack_groups": [
          {
            "name": "Initiative name",
            "summary": "Executive summary of this coordinated effort",
            "insight_indices": [0, 2, 5],
            "idea_indices": [1, 3, 4],
            "theme_indices": [0],
            "combined_effort": "medium",
            "combined_impact": "high",
            "execution_order": [3, 1, 4],
            "dependencies": "What must be in place first",
            "risks": "Key risks to watch for",
            "success_metrics": ["Metric 1", "Metric 2"]
          }
        ]
      }

      Guidelines:
      - Create 2-5 attack groups based on natural clusters
      - Prioritize groups with high impact and manageable effort
      - Ensure ideas in a group logically belong together
      - Consider dependencies between ideas when setting execution order
      - Include measurable success criteria
    PROMPT

    AttackGroup = Struct.new(
      :name, :summary, :insights, :ideas, :themes,
      :combined_effort, :combined_impact, :execution_order,
      :dependencies, :risks, :success_metrics,
      keyword_init: true
    )

    def build(insights:, ideas:, themes:)
      return { groups: [] } if insights.empty? && ideas.empty?

      insights_array = insights.to_a
      ideas_array = ideas.to_a
      themes_array = themes.to_a

      result = analyze_components(insights_array, ideas_array, themes_array)

      if result[:error]
        Rails.logger.error("Attack group building failed: #{result[:error]}")
        return { groups: [], error: result[:error] }
      end

      groups = build_groups_from_result(result, insights_array, ideas_array, themes_array)

      { groups: groups, count: groups.count }
    end

    private

    def analyze_components(insights, ideas, themes)
      prompt = build_components_prompt(insights, ideas, themes)
      call_claude(prompt, system_prompt: SYSTEM_PROMPT, max_tokens: 8192)
    end

    def build_components_prompt(insights, ideas, themes)
      parts = ["Create coordinated attack groups from the following components:\n"]

      parts << "\n=== INSIGHTS (#{insights.count}) ==="
      insights.each_with_index do |insight, index|
        parts << "--- Insight ##{index} ---"
        parts << "Title: #{insight.title}"
        parts << "Type: #{insight.insight_type} | Severity: #{insight.severity}"
        parts << "Description: #{insight.description[0..200]}..."
        parts << ""
      end

      parts << "\n=== IDEAS (#{ideas.count}) ==="
      ideas.each_with_index do |idea, index|
        parts << "--- Idea ##{index} ---"
        parts << "Title: #{idea.title}"
        parts << "Type: #{idea.idea_type} | Effort: #{idea.effort_estimate} | Impact: #{idea.impact_estimate}"
        parts << "Description: #{idea.description[0..200]}..."
        parts << ""
      end

      parts << "\n=== THEMES (#{themes.count}) ==="
      themes.each_with_index do |theme, index|
        parts << "--- Theme ##{index} ---"
        parts << "Name: #{theme.name}"
        parts << "Priority: #{theme.priority_score}"
        parts << "Description: #{theme.description[0..200]}..." if theme.description
        parts << ""
      end

      parts.join("\n")
    end

    def build_groups_from_result(result, insights, ideas, themes)
      groups_data = result[:attack_groups] || []

      groups_data.map do |group_data|
        build_group(group_data, insights, ideas, themes)
      end.compact
    end

    def build_group(data, insights, ideas, themes)
      group_insights = (data[:insight_indices] || []).map { |i| insights[i] }.compact
      group_ideas = (data[:idea_indices] || []).map { |i| ideas[i] }.compact
      group_themes = (data[:theme_indices] || []).map { |i| themes[i] }.compact

      return nil if group_insights.empty? && group_ideas.empty?

      ordered_ideas = if data[:execution_order].present?
        data[:execution_order].map { |i| ideas[i] }.compact
      else
        group_ideas
      end

      AttackGroup.new(
        name: data[:name],
        summary: data[:summary],
        insights: group_insights,
        ideas: ordered_ideas,
        themes: group_themes,
        combined_effort: data[:combined_effort],
        combined_impact: data[:combined_impact],
        execution_order: data[:execution_order],
        dependencies: data[:dependencies],
        risks: data[:risks],
        success_metrics: data[:success_metrics] || []
      )
    end
  end
end

File: apps/worker/src/ai/attack-group-builder.ts
/**
 * AI attack group builder — ports Rails AttackGroupBuilder.
 * Creates coordinated action bundles from insights, ideas, and themes.
 * Attack groups are ephemeral analysis results stored in pulse/reporting context.
 */
import { eq, desc } from "drizzle-orm";
import type { Database } from "@customer-pulse/db/client";
import { insights, ideas, themes } from "@customer-pulse/db/client";
import { callClaudeJson, resolveApiKey } from "./call-claude.js";

export interface AttackGroup {
  name: string;
  summary: string;
  idea_ids: number[];
  execution_order: string[];
  dependencies: string[];
  combined_effort: string;
  combined_impact: string;
  risks: string;
  success_metrics: string[];
}

const SYSTEM_PROMPT = `You are a product strategist creating coordinated action groups. Bundle related issues and solutions into 2-5 attack groups — each group is a coordinated set of actions that should be executed together for maximum impact.

For each attack group, return:
- "name": short name for the group (< 60 chars)
- "summary": 2-3 sentence executive summary
- "idea_ids": array of idea IDs to include in this group
- "execution_order": ordered array of step descriptions
- "dependencies": array of things that must be true before starting
- "combined_effort": one of "trivial", "small", "medium", "large", "extra_large"
- "combined_impact": one of "minimal", "low", "moderate", "high", "transformational"
- "risks": potential risks of executing this group
- "success_metrics": array of measurable success criteria

Return a JSON array of attack group objects.
Respond with ONLY the JSON array, no markdown fences.`;

export async function buildAttackGroups(
  db: Database,
  projectId: number,
): Promise<AttackGroup[]> {
  const apiKey = await resolveApiKey();
  if (!apiKey) return [];

  const recentInsights = await db
    .select({ id: insights.id, title: insights.title, description: insights.description, severity: insights.severity })
    .from(insights)
    .where(eq(insights.projectId, projectId))
    .orderBy(desc(insights.createdAt))
    .limit(20);

  const recentIdeas = await db
    .select({ id: ideas.id, title: ideas.title, description: ideas.description, effortEstimate: ideas.effortEstimate, impactEstimate: ideas.impactEstimate })
    .from(ideas)
    .where(eq(ideas.projectId, projectId))
    .orderBy(desc(ideas.createdAt))
    .limit(20);

  const recentThemes = await db
    .select({ id: themes.id, name: themes.name, description: themes.description, priorityScore: themes.priorityScore })
    .from(themes)
    .where(eq(themes.projectId, projectId))
    .orderBy(desc(themes.priorityScore))
    .limit(10);

  if (recentInsights.length === 0 && recentIdeas.length === 0) return [];

  const severityNames: Record<number, string> = { 0: "informational", 1: "minor", 2: "moderate", 3: "major", 4: "critical" };
  const effortNames: Record<number, string> = { 0: "trivial", 1: "small", 2: "medium", 3: "large", 4: "extra_large" };
  const impactNames: Record<number, string> = { 0: "minimal", 1: "low", 2: "moderate", 3: "high", 4: "transformational" };

  const context = [
    "INSIGHTS:",
    ...recentInsights.map((i) => `[ID:${i.id}] [${severityNames[i.severity]}] ${i.title}: ${i.description.slice(0, 120)}`),
    "",
    "IDEAS:",
    ...recentIdeas.map((i) => `[ID:${i.id}] [${effortNames[i.effortEstimate]}/${impactNames[i.impactEstimate]}] ${i.title}: ${i.description.slice(0, 120)}`),
    "",
    "THEMES:",
    ...recentThemes.map((t) => `[priority:${t.priorityScore}] ${t.name}: ${t.description?.slice(0, 100) ?? ""}`),
  ].join("\n");

  const result = await callClaudeJson<AttackGroup[]>({
    system: SYSTEM_PROMPT,
    user: `Create attack groups from these components:\n\n${context}`,
    maxTokens: 8192,
  });

  return result ?? [];
}
